### PR TITLE
Python 3 compatibility + a bit of pylint

### PR DESCRIPTION
**Pull request by bqpd**
Having a separate python3 branch was going to / got messy.

This PR has the necessary python3 changes (mostly cosmetic: maybe a few more unneeded objects created in some py2 operations) and pylint changes (totally cosmetic: caps in testing constants).

In this branch the unit test _sometimes_ all succeed; python 3 has randomized dictionary ordering, which screws with our tests quite a bit (though it doesn't hurt our actual functionality). The only solver curently supported in python3 is `mosek_cli`.

Don't trust github's diffs; pull this and do `git diff master`

Supersedes #206.

### CODE REVIEW

**Comment by bqpd** [comment]
@whoburg, if this looks good I'll merge it with master by hand.

**Comment by bqpd** [comment]
In this branch the unit test _sometimes_ all succeed; python 3 has randomized dictionary ordering, which screws with our tests quite a bit (though it doesn't hurt our actual functionality). The only solver curently supported in python3 is `mosek_cli`.

File: gpkit/geometric_program.py [hunk 1]
```diff
@@ -415,9 +415,10 @@ def _solve_sweep(self, printing, skipfailures, allownonoptimal):
         self.presweep = self.last
 
         if len(self.sweep) == 1:
-            sweep_grids = np.array(self.sweep.values())
+            sweep_grids = np.array(list(self.sweep.values()))
```
Comment (whoburg) [inline gpkit/geometric_program.py]: weird that a list is needed here -- how come?

File: gpkit/geometric_program.py [hunk 2]
```diff
@@ -486,7 +487,7 @@ def _parse_result(self, result, unsubbedexps, unsubbedvarlocs, cs, p_idxs,
 
         if result['status'] not in ["optimal", "OPTIMAL"]:
             if allownonoptimal:
-                print "It don't work! Letting it pass. This time."
+                print("It don't work! Letting it pass. This time.")
```
Comment (whoburg) [inline gpkit/geometric_program.py]: We should clean up this message.

File: gpkit/nomials.py [hunk 3]
```diff
@@ -17,6 +16,9 @@
 Quantity = ureg.Quantity
 Numbers += (Quantity,)
 
+def varkeysort(exp_item):
+    return list(exp_item[0].descr.items())
+
```
Comment (whoburg) [inline gpkit/nomials.py]: needs documentation -- I'm confused about where sorting is happening here

File: gpkit/nomials.py [hunk 4]
```diff
@@ -45,14 +47,14 @@ def __init__(self, exps=None, cs=1, varlocsandkeys=None,
...
-           and isinstance(exps, Strings + (VarKey, NoneType, dict))):
+           and (exps is None or isinstance(exps, Strings + (VarKey, dict)))):
             # building a Monomial
             if isinstance(exps, VarKey):
                 exp = {exps: 1}
                 units = exps.descr["units"] if "units" in exps.descr else None
             elif exps is None or isinstance(exps, Strings):
                 exp = {VarKey(exps, **descr): 1}
-                descr = exp.keys()[0].descr
+                descr = list(exp)[0].descr
```
Comment (whoburg) [inline gpkit/nomials.py]: creating a list here seems potentially expensive / unnecessary?

Comment (whoburg) [inline gpkit/nomials.py]: Oh, I guess it has length 1. But `list(exp)[0]` just seems ugly...

File: gpkit/tests/run_tests.py [hunk 5]
```diff
@@ -1,53 +1,56 @@
...
 
-import t_geometric_program
-tests += t_geometric_program.tests
+from .t_geometric_program import _TESTS as t_geometric_program
+_TESTS += t_geometric_program
 
-import t_gp_solution_array
-tests += t_gp_solution_array.tests
+from .t_gp_solution_array import _TESTS as t_gp_solution_array
+_TESTS += t_gp_solution_array
```
Comment (whoburg) [inline gpkit/tests/run_tests.py]: I think these should all be `TESTS`, not `_TESTS`. They're not so private/protected if we expect to import them in `run_tests`. 

See also 16719cb29c686154b30169cbf85830f3de4b58e4.

File: gpkit/tests/t_geometric_program.py [hunk 6]
```diff
@@ -72,11 +72,11 @@ def simpleflight_test_core(self, gp):
             self.assertTrue(abs(1-sol_rat) < 1e-2)
 
     def test_simpleflight(self):
-        from simpleflight import simpleflight_generator
+        from .simpleflight import simpleflight_generator
         sf = simpleflight_generator()
         self.simpleflight_test_core(sf.gp())
 
-    def test_Mddtest(self):
+    def tesT_Mddtest(self):
```
Comment (whoburg) [inline gpkit/tests/t_geometric_program.py]: why the cap T in `test`here?

File: gpkit/tests/t_geometric_program.py [hunk 7]
```diff
@@ -91,9 +91,15 @@ def test_Mddtest(self):
         self.assertEqual(gp2.A, CootMatrix(row=[0, 1],
                                            col=[0, 0],
                                            data=[-1, 1]))
-        self.assertEqual(gp3.A, CootMatrix(row=[0, 2, 3],
-                                           col=[0, 0, 0],
-                                           data=[-1, 1, -1]))
+        # order of variables with a posynomial is not stable
+        #   (though monomial order is)
```
Comment (whoburg) [inline gpkit/tests/t_geometric_program.py]: This makes me wonder whether we need a semi-convenient way to get the A matrix corresponding to a certain variable (column) ordering, e.g. `gp.A(vars=('x', 'y', 'z'))`. Or is that silly?

Is the order of monomial terms (rows/constraints) stable?

File: gpkit/tests/t_nomials.py [hunk 8]
```diff
@@ -194,15 +194,26 @@ def test_simplification(self):
         y = Monomial('y')
         p1 = x + y + y + (x+y) + (y+x**2) + 3*x
         p2 = 4*y + x**2 + 5*x
+        ps1 = [list(exp.keys())for exp in p1.exps]
+        ps2 = [list(exp.keys())for exp in p2.exps]
+        #print("%s, %s" % (ps1, ps2))  # python 3 dict reordering
```
Comment (whoburg) [inline gpkit/tests/t_nomials.py]: is this just debugging? `ps1` and `ps2` don't seem to get used in the test

File: gpkit/tests/t_nomials.py [hunk 9]
```diff
@@ -194,15 +194,26 @@ def test_simplification(self):
...
+        self.assertEqual(p1, p2)
+        p1 = (x+y)*(2*x+y**2)
         p2 = 2*x**2 + 2*y*x + y**2*x + y**3
-        self.assertEqual((x+y)*(2*x+y**2), p2)
+        ps1 = [list(exp.keys())for exp in p1.exps]
+        ps2 = [list(exp.keys())for exp in p2.exps]
+        #print("%s, %s" % (ps1, ps2))  # python 3 dict reordering
+        self.assertEqual(p1, p2)
```
Comment (whoburg) [inline gpkit/tests/t_nomials.py]: ditto above

**Comment by whoburg** [comment]
Before merging:

[ ] Added inline comments to address
[ ] Please try to merge the two commits 6feb52522475a2f064aa80169b8369cf27d8eb1d and 16719cb29c686154b30169cbf85830f3de4b58e4 currently living on the "cosmetics" branch (sorry for not having turned this into a pull request yet). Not sure on the best way here -- either merging those into this, or merging those into master first...
[ ] Unit tests should all pass on 2.7 (understood there are still issues with unit tests on python3)

File: gpkit/geometric_program.py (hunk 1 above)
Comment (bqpd) [inline gpkit/geometric_program.py]: in py3, `.values()` returns a `dictvalues` object, which is iterable but not a list.

File: gpkit/geometric_program.py (hunk 2 above)
Comment (bqpd) [inline gpkit/geometric_program.py]: Yeah, I think I wrote that just before our last meeting...

File: gpkit/tests/t_nomials.py (hunk 8 above)
Comment (whoburg) [inline gpkit/tests/t_nomials.py]: Also, space before `for`

File: gpkit/nomials.py (hunk 3 above)
Comment (bqpd) [inline gpkit/nomials.py]: Added, will push when I've responded to all your notes...

File: gpkit/nomials.py (hunk 4 above)
Comment (bqpd) [inline gpkit/nomials.py]: `exp.keys()` also generates a list dynamically. But, agreed. We don't have a great way to turn an exp into a varkey.

File: gpkit/tests/run_tests.py (hunk 5 above)
Comment (bqpd) [inline gpkit/tests/run_tests.py]: That looks great! Should I merge it in with this?

File: gpkit/tests/t_geometric_program.py (hunk 6 above)
Comment (bqpd) [inline gpkit/tests/t_geometric_program.py]: regex whoops

File: gpkit/geometric_program.py (hunk 1 above)
Comment (whoburg) [inline gpkit/geometric_program.py]: There's `np.fromiter`

More info: http://stackoverflow.com/questions/17371059/numpy-array-memory-allocation

File: gpkit/tests/t_geometric_program.py (hunk 7 above)
Comment (bqpd) [inline gpkit/tests/t_geometric_program.py]: Hmm, we could make CootMatrix accept those arguments.

Monomial terms are stable, yeah.

File: gpkit/tests/t_nomials.py (hunk 8 above)
Comment (bqpd) [inline gpkit/tests/t_nomials.py]: Oh, yeah. They're for the comment below them.

File: gpkit/tests/run_tests.py (hunk 5 above)
Comment (whoburg) [inline gpkit/tests/run_tests.py]: That would be great, along with another commit -- see checkbox in my summary comment below

File: gpkit/tests/t_geometric_program.py (hunk 7 above)
Comment (whoburg) [inline gpkit/tests/t_geometric_program.py]: Hmm, I like how lightweight CootMatrix is right now -- no dependency on gpkit objects. Vote for keeping that.

File: gpkit/geometric_program.py (hunk 1 above)
Comment (bqpd) [inline gpkit/geometric_program.py]: Profiled, this takes ~0% of the solve time, so I'm tempted to leave it be for readability.

Comment (whoburg) [inline gpkit/geometric_program.py]: fair enough.

**Pull request merged**